src/chips/safe_accumulator.rs

```diff
@@ -0,0 +1,211 @@
+use halo2_proofs::{circuit::*, halo2curves::pasta::Fp, plonk::*, poly::Rotation};
+
+#[derive(Debug, Clone)]
+pub struct SafeAccumulatorConfig<const MAX_BITS: u8, const ACC_COLS: usize> {
+    pub update_value: Column<Advice>,
+    pub left_most_inv: Column<Advice>,
+    pub add_carries: [Column<Advice>; ACC_COLS],
+    pub accumulate: [Column<Advice>; ACC_COLS],
+    pub instance: Column<Instance>,
+    pub is_zero: IsZeroConfig,
+    pub selector: [Selector; 2],
+}
+
+#[derive(Debug, Clone)]
+pub struct SafeACcumulatorChip<const MAX_BITS: u8, const ACC_COLS: usize> {
+    config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>,
+}
+
+impl<const MAX_BITS: u8, const ACC_COLS: usize> SafeACcumulatorChip<MAX_BITS, ACC_COLS> {
+    pub fn construct(config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>) -> Self {
+        Self { config }
+    }
+
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        update_value: Column<Advice>,
+        left_most_inv: Column<Advice>,
+        add_carries: [Column<Advice>; ACC_COLS],
+        accumulate: [Column<Advice>; ACC_COLS],
+        selector: [Selector; 2],
+        instance: Column<Instance>,
+    ) -> SafeAccumulatorConfig<MAX_BITS, ACC_COLS> {
+        let add_carry_selector = selector[0];
+        let overflow_check_selector = selector[1];
+
+        let is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(overflow_check_selector),
+            |meta| meta.query_advice(accumulate[0], Rotation::cur()),
+            left_most_inv,
+        );
+
+        // Enable equality on the advice and instance column to enable permutation check
+        meta.enable_equality(instance);
+        meta.enable_equality(update_value);
+        meta.enable_equality(left_most_inv);
+        accumulate.map(|col| meta.enable_equality(col));
+        add_carries.map(|col| meta.enable_equality(col));
+
+        meta.create_gate("accumulation constraint", |meta| {
+            let s_add = meta.query_selector(add_carry_selector);
+            let s_over = meta.query_selector(overflow_check_selector);
+
+            // value to be added
+            let value = meta.query_advice(update_value, Rotation::cur());
+
+            let previous_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(accumulate[i], Rotation::prev()))
+                .collect::<Vec<Expression<Fp>>>();
+            let carries_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(add_carries[i], Rotation::cur()))
+                .collect::<Vec<Expression<Fp>>>();
+            let updated_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(accumulate[i], Rotation::cur()))
+                .collect::<Vec<Expression<Fp>>>();
+
+            let shift_next_chunk = Expression::Constant(Fp::from(1 << MAX_BITS));
+
+            // Add the value with right most accumulation column
+            let check_add_value_exprs = vec![
```

Comment:
Can you better explain this constraint?

Reply:
For example, let's assume that each accumulate column has a maximum value of 0xf.

If we add the value 0x5 to the decomposition of `previous_acc` which has the values `0xed` in the `accumulate_1` and `accumulate_0` columns.

| -             | new_value | left_most_inv | add_carries_2 | add_carries_1 | add_carries_0 | accumulate_2 | accumulate_1 | accumulate_0 | add_selector |
| --            | --        | --            | --            | --            | --            | --           | --           | -            | -            |
| previous_acc  |           |               |               |               |               | 0            | 0xe          | → 0xd        | 0            | 
| updated_acc   | → 0x5     | 0             | 0             | 0             | 1 ←           | 0            | 0xf          | 2   ←        | 1            |

In this case, the result is that `0xd + 0x5 = 0x12`.
The first digit of the result, '0x12', is 0x1, which is assigned to the `add_carries_0` column, and the rest of the result, 0x2, is placed in the `accumulate_0` column.
In other words, the sum of the value between the rightmost value of the `accumulate` columns and the value of `new_value` is divided into the carry and the remainder.

Note that `←` and `→` are used to distinguish between addition and subtraction in the constraints.

Can express the constraints like this:
`( new_value[1] + accumulate_0[0] ) - ( accumulate_0[1] + ( add_carries_1[1] * 16 ) ) = 0`

Reply:
I'm reading back the [Add carry V1 Chip](https://github.com/summa-dev/halo2-experiments/tree/overflow-check#experiment-9---add-carry-v1). Wouldn't the constraint performed there accomplish the same thing? 

The core goal of the chip is to:
- Take in a set of values
- Iteratively add them to the accumulator

The constraints are:
- Make sure that the addition is performed correctly every round
- Make sure that the accumulator doesn't overflow

Let me try to apply the `V1` chip here, considering each column as an accumulator of 4 bits.

| -   | value      | acc_1 | acc_0 |
| --- | ---------- | ----- |:----- |
| 0   | -          | 0     | 0     |
| 1   | 237 (0xed) | 0xe   | 0xd   |
| 2   | 5    (0x5) | 0xf   | 0x2   |

The constraints here can be: 
- `acc[1](prev) << 4 + acc[0](prev) + value(cur) - acc[1](cur) << 4 - acc[0](cur) = 0` 
- Check that every value inside an `acc` column is within a BYTE using a 4_BITS range check

Note that adding value that causes the accumulator to overflow, in this case, 8 bits won't ever satisfy the first constraint. 

| -   | value       | acc_1 | acc_0 |
| --- | ----------- | ----- |:----- |
| 0   | -           | 0     | 0     |
| 1   | 237 (0xed)  | 0xe   | 0xd   |
| 2   | 19   (0x13) | 0x0   | 0x0   |

I'm trying to understand if there's any crucial functionality performed by the `safe_accumulator` chip that is not performed here

Reply:
> I'm reading back the [Add carry V1 Chip](https://github.com/summa-dev/halo2-experiments/tree/overflow-check?rgh-link-date=2023-05-02T14%3A28%3A22Z#experiment-9---add-carry-v1). Wouldn't the constraint performed there accomplish the same thing?

I think it's similar, but not exactly the same thing. In the `add_carry_v1` chip, all columns, `hi` and `lo`, are used for constraints. However, here, only the right-most accumulation column is used with the `new_value` for creating constraints.

> | - | value | acc_1 | acc_0
> | -- | -- | -- | --
> | 0 | - | 0 | 0
> | 1 | 237 (0xed) | 0xe | 0xd
> | 2 | 19 (0x13) | 0x0 | 0x0
>
> I'm reading back the [Add carry V1 Chip](https://github.com/summa-dev/halo2-experiments/tree/overflow-check?rgh-link-date=2023-05-02T14%3A28%3A22Z#experiment-9---add-carry-v1). Wouldn't the constraint performed there accomplish the same thing?

That's why always we need one more extra column, like in [here - overflow_check.rs](https://github.com/summa-dev/halo2-experiments/tree/overflow-check#experiment-11---overflow-check). 

For your example, we should add one more extra column as shown below. This is also the reason why we cannot use the `add carry` chip in the `safe_accumulator`.

| - | value | acc_2 | acc_1 | acc_0
| -- | -- | -- | -- | --
| 0 | - | - | 0 | 0
| 1 | 237 (0xed) | 0x0 |  0xe | 0xd
| 2 | 19 (0x13) | 0x1 | 0x0 | 0x0

FYI, the maximum value of the new_value column is also limited, just like each accumulation column. We should work on improving this by decomposing the `new_value` within the circuit.

Reply:
I'm still not clear about the advantages of the `safe_accumulator` chip over the `add_carry_v1`. Is it efficiency? or there are some constraints that were missing the `add_carry_v1` that were added in the `safe_accumulator`?

If that's not the case, I would privilege the `add_carry_v1` chip, just for the overall simplicity of the chip and readability of the code.

> For your example, we should add one more extra column as shown below.

Do you mean that we should the column in order to check that an overflow happened? If we are checking that the accumulated value doesn't overflow 8-bits, it seems that 2 columns are enough (with the constraints set as mentioned in my previous comment)

Reply:
> I'm still not clear about the advantages of the `safe_accumulator` chip over the `add_carry_v1`. Is it efficiency? or there are some constraints that were missing the `add_carry_v1` that were added in the `safe_accumulator`?
> 

I understand your point. Using the `add_carry_v1` chip as an accumulator might be more efficient than the safe_accumulator. However, I am not certain about its performance with large numbers, the current `add_carry` implementation has an iteration for counting if more than `max_bits` in [here - utils.rs f_to_nbits](https://github.com/summa-dev/halo2-experiments/blob/aa268f58c974004802af2b33872e4edbef0abe4a/src/chips/utils.rs#L42-L45) 

 However, one limitation of the `add_carry_v1` chip is that it cannot handle numbers larger than the modular limit of the finite fields in the circuit. This is an advantage that the `safe_accumulator` chip has over the `add_carry_v1` chip.

> If that's not the case, I would privilege the `add_carry_v1` chip, just for the overall simplicity of the chip and readability of the code.

In our discussion about the maximum possible value in Summa, I think that using `add_carry_v1` as an accumulator would be appropriate if we only consider values up to 2^252, which is less than the field modulus. 

However, this implementation is an experiment based on the assumption that we may need to use the `safe_accumulator` in the future if we have to deal with numbers larger than the field modulus.

> > For your example, we should add one more extra column as shown below.
> 
> Do you mean that we should add the column in order to check that an overflow happened? If we are checking that the accumulated value doesn't overflow 8-bits, it seems that 2 columns are enough (with the constraints set as mentioned in my previous comment)

I believe my previous response may have been incorrect.
I took the data in your second example( 0x13 + 0xed - 0x0 ) and assumed it was on the `safe_accumulator' chip (by checking the leftmost accumulator column).
 You're right, two columns should be sufficient in this case.

Reply:
> However, one limitation of the `add_carry_v1` chip is that it cannot handle numbers larger than the modular limit of the finite fields in the circuit. This is an advantage that the `safe_accumulator` chip has over the `add_carry_v1` chip.
> 
> >

I totally see your point here. Let me copy the example you shared with me off-github to show the problem of the `add_carry_v1` chip.  Considering a case where the finite field modulus is `255` and the following witness table, the constraint `( 0x13 + ( 0xe * ( 1 << 4 ) + 0xd) - 0x0 == 0 )` will be satisfied, which is a bug! 

| -   | value      | acc_1 | acc_0 |
| --- | ---------- | ----- |:----- |
| 0   | -          | 0     | 0     |
| 1   | 237 (0xed) | 0xe   | 0xd   |
| 2   | 19   (0x13)| 0     |  0    |

Since the `safe_accumulator` would be able to handle an accumulator larger than the field modulus, which is a desired feature, let's move forward with that implementation. I'll continue my code review focusing on that chip!

Reply:
I was looking again into this constraint. 

| new_value | left_most_inv | add_carries_2 | add_carries_1 | add_carries_0 | accumulate_2 | accumulate_1 | accumulate_0 |
| --------- | ------------- | ------------- | ------------- | ------------- |:------------ |:------------ |:------------ |
|           |               |               |               |               |              | 0xe          | 0xd          |
| 0x64      |               |               | 1             | 1             | 0x1          | 0x5          | 0x1          |

In this case, it seems that the constraint
`( new_value[1] + accumulate_0[0] ) - ( accumulate_0[1] + ( add_carries_1[1] * 16 ) ) = 0` won't be satisfied 

as it would be 

= `0x64 + 0xd - (0x1 + 0x1  * 16)`

Is that correct?

Furthermore I see other 2 missing constraints here:

- as mentioned before, checking the value inside each `accumulate` column is in `MAX_BITS` range
- check that the value inside `add_carries` is binary

Reply:
>In this case, it seems that the constraint `( new_value[1] + accumulate_0[0] ) - ( accumulate_0[1] + ( add_carries_1[1] * 16 ) ) = 0` won't be satisfied
> 
> as it would be
> 
> = `0x64 + 0xd - (0x1 + 0x1 * 16)`
> 
> Is that correct?

Yes, that is correct. and it's the reason for checking the value within the maximum using the assert term.
https://github.com/summa-dev/halo2-experiments/blob/16df4fc8cdcd83275ad9dcb2fa4af82c7c6301c0/src/chips/safe_accumulator.rs#L160-L163
I know it's not the right way, so I fixed with using range check instead using `assert` at https://github.com/summa-dev/halo2-experiments/commit/0ad2ed3c1a5eb59570bf0dfc53242c612bacdb05

Also, that is a limitation of this experiment. 
the value for adding to accumulates columns can be handled more than a larger number by adjusting `MAX_BITS`
Or we can add a feature that decomposes the values in cells for handling larger the value, like below in the  `overflow_check_v2` chip.
https://github.com/summa-dev/halo2-experiments/blob/0ad2ed3c1a5eb59570bf0dfc53242c612bacdb05/src/chips/overflow_check_v2.rs#L102-L107

Following this, we'd need to adjust the constraints to account for the chunk-wise sum

> 
> Furthermore I see other 2 missing constraints here:
> 
> * as mentioned before, checking the value inside each `accumulate` column is in `MAX_BITS` range

Checking values inside the accumulate column here
https://github.com/summa-dev/halo2-experiments/blob/0ad2ed3c1a5eb59570bf0dfc53242c612bacdb05/src/chips/safe_accumulator.rs#L144-L145
the util function `range_check_vec` generate expressions for checking accumulates values in the range

> * check that the value inside `add_carries` is binary

Ah! Yes, I think it needs.

How about this as constraint ` selector * carried_value * ( carried_value - 1)` ?

Reply:
Ok, thanks for the clarification. Although I recognize the limitation, I think that the chip is sound as it is and we can keep it as it is for now. 



> How about this as constraint ` selector * carried_value * ( carried_value - 1)` ?

Yes, sounds perfect!

Reply:
Updated at [13af60d](https://github.com/summa-dev/halo2-experiments/pull/5/commits/13af60d0d19f01dbffc56897c45ec7eacb7ff7ed)